File: packages/api-models-ts/src/Comments.ts
/**
 * Request and response models for source comments.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Groups from "./Groups.ts";
import * as Schemas from "./Schemas.ts";

/**
 * What a comment is attached to.
 *
 * @since 1.0.0
 * @category Models
 */
export type CommentResourceType =
    | "sources"
    | "spectra"
    | "gcn_event"
    | "shift"
    | "earthquake";

/**
 * The fields of a {@link Comment}.
 *
 * @since 1.0.0
 * @category Models
 */
export const CommentEntries = {
    id: Schemas.Integer,
    created_at: Schemas.NullishTimestamp,
    modified: Schemas.NullishTimestamp,
    text: Schemas.NullishString,
    attachment_name: Schemas.NullishString,
    attachment_bytes: Schemas.nullish(Schemas.Json),
    origin: Schemas.NullishString,
    bot: Schemas.NullishBoolean,
    author_id: Schemas.NullishInteger,
    author: Schemas.nullish(Schemas.JsonObject),
    groups: Schemas.nullish(v.array(Groups.Group)),
    obj_id: Schemas.NullishString,
    spectrum_id: Schemas.NullishInteger,
    gcn_id: Schemas.NullishInteger,
    earthquake_id: Schemas.NullishInteger,
    shift_id: Schemas.NullishInteger,
    obj: Schemas.nullish(Schemas.JsonObject),
    spectrum: Schemas.nullish(Schemas.JsonObject),
    gcn: Schemas.nullish(Schemas.JsonObject),
    shift: Schemas.nullish(Schemas.JsonObject),
    earthquake: Schemas.nullish(Schemas.JsonObject),
    dateobs: Schemas.NullishTimestamp,
    resourceType: Schemas.NullishString,
};

/**
 * A comment on any commentable resource (upstream `Comment`).
 *
 * Upstream splits comments across `Comment`, `CommentOnSpectrum`,
 * `CommentOnGCN`, `CommentOnShift` and `CommentOnEarthquake`; this model is
 * the union of that family, so each type-specific foreign key is optional and
 * only the ones belonging to the comment's own table are ever set. `author` is
 * the author's `User.to_dict()` (plus a `gravatar_url` key on the source
 * endpoints), and `obj`, `gcn`, `spectrum`, `shift` and `earthquake` stay
 * free-form to avoid importing in a circle from the modules that import this
 * one.
 *
 * @since 1.0.0
 * @category Models
 */
export const Comment = Schemas.model(v.strictObject(CommentEntries));

/**
 * @since 1.0.0
 * @category Models
 */
export type Comment = v.InferOutput<typeof Comment>;

/**
 * A single comment, as returned by the single-comment endpoint.
 *
 * The list and single-GET routes both return `Comment.to_dict()` plus
 * `resourceType`, so this is {@link Comment} under the name the single-comment
 * endpoint is documented with.
 *
 * @since 1.0.0
 * @category Models
 */
export const CommentDetail = Schemas.model(v.strictObject(CommentEntries));

/**
 * @since 1.0.0
 * @category Models
 */
export type CommentDetail = v.InferOutput<typeof CommentDetail>;

/**
 * Result of posting a comment.
 *
 * @since 1.0.0
 * @category Models
 */
export const CommentPostResponse = Schemas.model(
    v.strictObject({
        comment_id: Schemas.Integer,
        message: Schemas.NullishString,
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type CommentPostResponse = v.InferOutput<typeof CommentPostResponse>;

/**
 * The decoded contents of a comment attachment.
 *
 * @since 1.0.0
 * @category Models
 */
export const CommentAttachment = Schemas.model(
    v.strictObject({
        commentId: Schemas.Integer,
        attachment: Schemas.NullishString,
        attachmentName: Schemas.NullishString,
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type CommentAttachment = v.InferOutput<typeof CommentAttachment>;

/**
 * How many comments still hold their attachment in the database.
 *
 * @since 1.0.0
 * @category Models
 */
export const CommentAttachmentCounts = Schemas.model(
    v.strictObject({
        totalWithoutAttachmentBytes: v.optional(Schemas.Integer, 0),
        totalWithAttachmentBytes: v.optional(Schemas.Integer, 0),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type CommentAttachmentCounts = v.InferOutput<typeof CommentAttachmentCounts>;

/**
 * Result of moving one page of comment attachments to disk.
 *
 * @since 1.0.0
 * @category Models
 */
export const CommentAttachmentBatch = Schemas.model(
    v.strictObject({
        totalMatches: v.optional(Schemas.Integer, 0),
        pageNumber: v.optional(Schemas.Integer, 1),
        numPerPage: v.optional(Schemas.Integer, 100),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type CommentAttachmentBatch = v.InferOutput<typeof CommentAttachmentBatch>;

File: packages/api-models-py/skyportal_py_models/comments.py
"""Request and response models for source comments."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from skyportal_py_models.groups import Group


class Comment(BaseModel):
    """A comment on any commentable resource (upstream ``Comment``).

    Upstream splits comments across ``Comment``, ``CommentOnSpectrum``,
    ``CommentOnGCN``, ``CommentOnShift`` and ``CommentOnEarthquake``; this
    model is the union of that family, so each type-specific foreign key
    is optional and only the ones belonging to the comment's own table are
    ever set. ``author`` is the author's ``User.to_dict()`` (plus a
    ``gravatar_url`` key on the source endpoints), and ``obj``, ``gcn``,
    ``spectrum``, ``shift`` and ``earthquake`` stay ``dict`` to avoid
    importing in a circle from the modules that import this one.
    """

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    id: int
    created_at: datetime | None = None
    modified: datetime | None = None
    text: str | None = None
    attachment_name: str | None = None
    attachment_bytes: Any = None
    origin: str | None = None
    bot: bool | None = None
    author_id: int | None = None
    author: dict[str, Any] | None = None
    groups: list[Group] | None = None
    obj_id: str | None = None
    spectrum_id: int | None = None
    gcn_id: int | None = None
    earthquake_id: int | None = None
    shift_id: int | None = None
    obj: dict[str, Any] | None = None
    spectrum: dict[str, Any] | None = None
    gcn: dict[str, Any] | None = None
    shift: dict[str, Any] | None = None
    earthquake: dict[str, Any] | None = None
    dateobs: datetime | None = None
    resource_type: str | None = Field(alias="resourceType", default=None)


class CommentPostResponse(BaseModel):
    """Result of posting a comment."""

    model_config = ConfigDict(extra="forbid")

    comment_id: int
    message: str | None = None


class CommentDetail(Comment):
    """A single comment, as returned by the single-comment endpoint.

    The list and single-GET routes both return ``Comment.to_dict()`` plus
    ``resourceType``, so this is :class:`Comment` under the name the
    single-comment endpoint is documented with.
    """


class CommentAttachment(BaseModel):
    """The decoded contents of a comment attachment."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    comment_id: int = Field(alias="commentId")
    attachment: str | None = None
    attachment_name: str | None = Field(alias="attachmentName", default=None)


class CommentAttachmentCounts(BaseModel):
    """How many comments still hold their attachment in the database."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    total_without_attachment_bytes: int = Field(
        alias="totalWithoutAttachmentBytes", default=0
    )
    total_with_attachment_bytes: int = Field(
        alias="totalWithAttachmentBytes", default=0
    )


class CommentAttachmentBatch(BaseModel):
    """Result of moving one page of comment attachments to disk."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    total_matches: int = Field(alias="totalMatches", default=0)
    page_number: int = Field(alias="pageNumber", default=1)
    num_per_page: int = Field(alias="numPerPage", default=100)
